src/v/container/include/container/contiguous_range_map.h

```diff
@@ -0,0 +1,303 @@
+            auto tmp = *this;
+            --*this;
+            return tmp;
+        }
+
+        bool operator==(const iter& o) const {
+            return std::tie(_it, _container) == std::tie(o._it, o._container);
+        };
+
+        auto operator<=>(const iter& o) const {
+            return std::tie(_it, _container) <=> std::tie(o._it, o._container);
+        };
+
+    private:
+        void skip_to_next_present() {
+            while (_it != _container->_values.end()
+                   && !_container->_index_bitmap.contains(
+                     boost::numeric_cast<unsigned_t>(_it->first))) {
+                ++_it;
+            }
+        }
+
+        friend class contiguous_range_map<KeyT, ValueT>;
+        using parent_t = std::conditional_t<
+          Const,
+          const contiguous_range_map<KeyT, ValueT>,
+          contiguous_range_map<KeyT, ValueT>>;
+
+        iter(
+          parent_t* map,
+          typename parent_t::underlying_t::template iter<Const> it)
+          : _it(it)
+          , _container(map) {
+            skip_to_next_present();
+        }
+
+        typename parent_t::underlying_t::template iter<Const> _it;
+        parent_t* _container;
+    };
+
+public:
+    using iterator = iter<false>;
+    using const_iterator = iter<true>;
+
+    contiguous_range_map() noexcept = default;
+    contiguous_range_map& operator=(contiguous_range_map&& other) noexcept {
+        if (this != &other) {
+            this->_size = other._size;
+            this->_values = std::move(other._values);
+            other._size = 0;
+        }
+        return *this;
+    }
+
+    contiguous_range_map(contiguous_range_map&& other) noexcept {
+        *this = std::move(other);
+    }
+    contiguous_range_map(const contiguous_range_map&) = delete;
+    contiguous_range_map& operator=(const contiguous_range_map&) = delete;
+
+    ~contiguous_range_map() noexcept = default;
+
+    /**
+     * Determines if element comparing equal to given key exists in a map.
+     */
+    bool contains(KeyT key) const {
+        check_key(key);
+        return _index_bitmap.contains(boost::numeric_cast<unsigned_t>(key));
+    }
+
+    /**
+     * Returns an element comparing equal to given key. If an element for the
+     * given key is not present the default element will be constructed.
+     */
+    mapped_type& operator[](KeyT key) { return emplace(key).first->second; }
+    /**
+     * Constructs element for given key in place.
+     *
+     * Returns a pair containing an iterator to the element (either existing or
+     * newly created one) and a boolean indicating if element was inserted.
+     *
+     * Element constructor is not invoked if an element already exists in the
+     * map.
+     */
+    template<typename... Args>
+    std::pair<iterator, bool> emplace(KeyT key, Args&&... args) {
+        check_key(key);
+        auto current_key = _values.size();
+        while (static_cast<size_t>(key) >= _values.size()) {
```

Comment:
(feel free to ignore, not suggesting to implement in this PR).

one tiny potential optimization is when key == _values.size() (aka emplacing the next contiguous key, which is _probably_ the most common case?), we can have a fast path emplace that directly constructs with the args in the right slot rather than default constructing  first and then moving into it like we do here.

Reply:
good point, will do it in the follow up PR.